qtwebengine/src/core/content_browser_client_qt.cpp

```diff
@@ -217,24 +222,44 @@ class MessagePumpForUIQt : public QObject,
 
 private:
     bool handleScheduledWork() {
-        bool more_work_is_plausible = m_delegate->DoWork();
-
-        base::TimeTicks delayed_work_time;
-        more_work_is_plausible |= m_delegate->DoDelayedWork(&delayed_work_time);
-
-        if (more_work_is_plausible)
-            return true;
-
-        more_work_is_plausible |= m_delegate->DoIdleWork();
-        if (!more_work_is_plausible)
-            ScheduleDelayedWork(delayed_work_time);
-
-        return more_work_is_plausible;
+        unsigned int num_completed(0);
+        bool did_work(false);
+
+        while (num_completed < m_eventCount) {
+            num_completed++;
+
+            bool more_work_is_plausible = m_delegate->DoWork();
+
+            base::TimeTicks delayed_work_time;
+            more_work_is_plausible |= m_delegate->DoDelayedWork(&delayed_work_time);
+            did_work |= more_work_is_plausible;
+
+            // If there could be either more work or more delayed work continue
+            // in an attempt to do more.
+            if (more_work_is_plausible)
+            {
+                continue;
+            }
+
+            // Attempt to do idle work.
+            more_work_is_plausible |= m_delegate->DoIdleWork();
+            did_work |= more_work_is_plausible;
+
+            // If we didn't complete any idle work schedule next check for
+            // delayed work and return.
+            if (!more_work_is_plausible)
+            {
+                ScheduleDelayedWork(delayed_work_time);
+                return did_work;
```

Comment:
@war1025 I tried to create a "wrapper" function but there was not a good way to get this condition to exit the "outer loop" while returning the correct value without using exceptions. I'm completely open to any suggestions that you might have.